lib/algos/list/delimiterMaster.js

```diff
@@ -1,84 +1,112 @@
-            // In cases of maxKeys <= 0 => IsTruncated = false
-            this.IsTruncated = this.maxKeys > 0;
-            return false;
+        let key = obj.key;
+        const value = obj.value;
+        if ((this.prefix && !key.startsWith(this.prefix))
+                || (typeof this.NextMarker === 'string' &&
+                    key <= this.NextMarker)) {
+            return FILTER_SKIP;
         }
-        // <versioning>
-        const value = VSUtils.decodeVersion(obj.value);
-        // ignore it if the master version is a delete marker
-        if (VSUtils.isDeleteMarker(value)) {
-            return true;
+        const versionIdIndex = key.indexOf(VID_SEP);
+        if (versionIdIndex >= 0) {
+            // generally we do not accept a specific version,
+            // we only do when the master version is a PHD version
+            key = key.slice(0, versionIdIndex);
+            if (key !== this.prvPHDKey) {
+                return FILTER_ACCEPT; // trick repd to not increase its streak
+            }
+        }
+        if (Version.isPHD(value)) {
+            // master version is a PHD version: wait for the next version
+            this.prvPHDKey = key;
+            return FILTER_ACCEPT; // trick repd to not increase its streak
+        }
+        if (Version.isDeleteMarker(value)) {
+            // version is a delete marker, ignore
+            return FILTER_ACCEPT; // trick repd to not increase its streak
         }
-        // use the original object name for delimitering to work correctly
-        const key = VSUtils.getObjectNameFromMasterKey(obj.key);
-        // </versioning>
+        // non-PHD master version or a version whose master is a PHD version
+        this.prvPHDKey = undefined;
         if (this.delimiter) {
-            const commonPrefixIndex =
-                key.indexOf(this.delimiter, this.searchStart);
-            if (commonPrefixIndex === -1) {
-                this.addContents(key, value);
-            } else {
-                this.addCommonPrefix(
-                    key.substring(0, commonPrefixIndex + this.delimLen));
+            // check if the key has the delimiter
+            const baseIndex = this.prefix ? this.prefix.length : 0;
+            const delimiterIndex = key.indexOf(this.delimiter, baseIndex);
+            if (delimiterIndex < 0) {
+                return this.addContents(key, value);
             }
-        } else {
-            this.addContents(key, value);
+            // try to add the prefix to the list
+            return this.addCommonPrefix(key, delimiterIndex);
+        }
+        return this.addContents(key, value);
+    }
+
+    skipping() {
+        if (this.NextMarker) {
+            // next marker:
+            // - foo/ : skipping foo/
+            // - foo  : skipping foo.
+            const index = this.NextMarker.lastIndexOf(this.delimiter);
+            return this.NextMarker +
+                (index === this.NextMarker.length ? '' : VID_SEP);
```

Comment:
We can use that but then we need to switch the order of the results. The purpose is to use `VID_SEP` as a sub-delimiter in the case we need to skip a range of versions of the same object (`this.NextMarker` is not delimited by the main delimiter). I have the comments above stating that clearly.